servo: Add an `Info.plist` on the Mac and opt into integrated graphics.

pcwalton opened the pull request:
Discrete GPUs cause power use problems and tend to perform badly
with WebRender.

See:
- https://developer.apple.com/library/mac/qa/qa1734/_index.html
- https://reverse.put.as/2013/05/28/gimmedebugah-how-to-embedded-a-info-plist-into-arbitrary-binaries/

r? @metajack

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/servo/servo/11646)

<!-- Reviewable:end -->

pcwalton committed:
servo: Add an `Info.plist` on the Mac and opt into integrated graphics.

Discrete GPUs cause power use problems and tend to perform badly with
WebRender.

See:

* https://developer.apple.com/library/mac/qa/qa1734/_index.html

* https://reverse.put.as/2013/05/28/gimmedebugah-how-to-embedded-a-info-plist-into-arbitrary-binaries/

highfive commented:
<img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20> **Warning** <img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20>

* These commits modify **unsafe code**. Please review it carefully!

highfive commented:
New code was committed to pull request.

KiChjang commented:
Fails tidy:
```
103.74s$ ./mach test-tidy --no-progress --all

Checking files for tidiness...

./components/servo/info_plist.rs:10: missing space after ,

./components/servo/info_plist.rs:10: missing space before =

./components/servo/info_plist.rs:10: missing space after =

./components/servo/info_plist.rs:12: missing space before -

./components/servo/info_plist.rs:16: missing space after /

./components/servo/info_plist.rs:17: missing space after /

./components/servo/info_plist.rs:18: missing space after /

./components/servo/info_plist.rs:19: missing space after /

./components/servo/info_plist.rs:20: missing space after /

./components/servo/info_plist.rs:21: missing space after /

./components/servo/info_plist.rs:22: missing space after /

./components/servo/info_plist.rs:23: missing space before /

./components/servo/info_plist.rs:23: missing space after /

./components/servo/info_plist.rs:23: missing space after /

./components/servo/info_plist.rs:23: missing space after /

./components/servo/info_plist.rs:24: missing space after /

./components/servo/info_plist.rs:25: missing space after /

./components/servo/info_plist.rs:26: missing space after /

./components/servo/info_plist.rs:27: missing space after /

./components/servo/info_plist.rs:28: missing space after /

./components/servo/info_plist.rs:29: missing space after /

./components/servo/info_plist.rs:30: missing space after /

./components/servo/info_plist.rs:31: missing space after /
```

metajack commented:
@pcwalton Is this integrated with the work on `./mach package`

@edunham might want to look at this too.

paulrouget commented on components/servo/info_plist.rs:
```diff
@@ -0,0 +1,44 @@
+       <key>CFBundlePackageType</key>
+       <string>APPL</string>
+       <key>CFBundleIconFile</key>
+       <string>../../resources/servo.png</string>
```
This might not always work. Can you use `resources_dir_path()` for this?

pcwalton replied:
Oh, oops. This is an old version of the patch. Will fix ASAP. Do not merge.

jdm commented on components/servo/main.rs:
```diff
@@ -97,6 +101,8 @@ fn unregister_glutin_resize_handler(window: &Rc<app::window::Window>) {
     }
 }
 
+// Just as the name implies, this is a hack to prevent the linker from stripping our `info.plist`.
```
Weird place for this comment.

highfive commented:
New code was committed to pull request.

pcwalton committed:
mach: Set a nice icon for the binary on the Mac.

There's actually a practical reason for this: it helps me pick out the
app in Instruments.app.

See:
http://apple.stackexchange.com/questions/6901/how-can-i-change-a-file-or-folder-icon-using-the-terminal/161984#161984

highfive commented:
New code was committed to pull request.

pcwalton commented:
@metajack Updated. r?

I believe your `Info.plist` that you place in the `.app` bundle will override the one here, so it shouldn't cause any problems.

metajack commented:
There are several problems here:

- The code unconditionally imports Cocoa, which may cause mach to fail and which is not documented. This should be optional just like our build length notification.
- It's not clear why you are adding info_plist.rs to the ignores for tidy.
- I don't want to see `info_plist::hack_to_prevent_linker_from_stripping_the_info_plist()` in `main()`. Can we hide this better? Maybe we need a `platform::init()` where this kind of stuff can go.
- The Info.plist file needs to be stored somewhere in tree and included `with include_str!`, so that it can be reused by the `./mach package` command. I don't mind it being linked in and on disk in a final package if they are the same, but we should at least not have two copies around or have to extract things from a build artifact.

On another note, is there seriously no way to tell rust not to optimize away a function? That is such an ugly hack that we should make sure a rust bug is filed for this and have a // TODO over that function that references that bug so that we know the criteria for when we can clean it up later.

jdm commented:
There's [test::black_box](https://github.com/rust-lang/rust/blob/0ec321f7b541fcbfbf20286beb497e6d9d3352b2/src/libtest/lib.rs#L1196).

pcwalton commented:
> It's not clear why you are adding info_plist.rs to the ignores for tidy.

This is because tidy doesn't realize that `__TEXT,__info_plist` is inside a quoted string and will fail telling me to insert a space after the comma. That breaks because Mach-O (I believe) requires no space after the comma.

metajack commented:
@pcwalton Fair, but let's file a bug against tidy and annotate the reason in the ignores list.

edunham commented:
@metajack `mach package` currently only supports Linux and Android. 

I haven't worked with OSX packaging much, but I don't see anything obviously wrong in these changes.

pcwalton commented:
@jdm Sadly `test::black_box` only [prevents LLVM's optimizer from optimizing out the symbol](https://doc.rust-lang.org/1.1.0/src/test/lib.rs.html#1054-1059), which doesn't help us in this case because it's the *linker* that strips out the symbol.

pcwalton committed:
servo: Update `cocoa-rs` to pick up servo/cocoa-rs#131.

highfive commented:
New code was committed to pull request.

pcwalton commented:
Comments addressed.

r? @metajack

highfive commented:
New code was committed to pull request.

highfive commented:
New code was committed to pull request.

metajack commented:
Reviewable has totally failed me on this patch. Sigh.

metajack commented:
@bors-servo r+

Thanks for these updates.

bors-servo commented:
:pushpin: Commit 8d5ff5c has been approved by `metajack`

<!-- @bors-servo r=metajack 8d5ff5cae765ff9f7711c564685ebb15a52ee96f -->

bors-servo commented:
:hourglass: Testing commit 8d5ff5c with merge e51ce5d...

bors-servo commented:
:sunny: Test successful - [android](http://build.servo.org/builders/android/builds/1364), [arm32](http://build.servo.org/builders/arm32/builds/1456), [arm64](http://build.servo.org/builders/arm64/builds/1443), [linux-dev](http://build.servo.org/builders/linux-dev/builds/1468), [linux-rel](http://build.servo.org/builders/linux-rel/builds/1533), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/1510), [mac-rel-css](http://build.servo.org/builders/mac-rel-css/builds/1495), [mac-rel-wpt](http://build.servo.org/builders/mac-rel-wpt/builds/1551), [windows](http://build.servo.org/builders/windows/builds/1177)
